src/Aspire.Hosting.Azure/ProvisioningExtensions.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Publishing;
+using Microsoft.Extensions.Configuration;
+
+namespace Aspire.Hosting.Azure;
+internal static class ProvisioningExtensions
+{
+    internal static void WriteBicepResourceToManifest(ManifestPublishingContext context, IResource resource)
+    {
+        context.Writer.WriteString("type", "azure.bicep.v0");
+
+        // Grab environment variables that would be passed in from AZD.
+        var tenantId = context.Configuration.GetValue<Guid>("Azure:TenantId");
+        var subscriptionId = context.Configuration.GetValue<Guid>("Azure:SubscriptionId");
+        var environmentName = context.Configuration.GetValue<string>("Azure:EnvironmentName") ?? throw new DistributedApplicationException("Boom!");
+
+        // This is our implementation of an IConstruct (subscription level) to which resources can be added.
+        var bicepGenerationContext = new BicepGenerationContext(context.AppModel, resource, tenantId, subscriptionId,environmentName);
+
+        // All resources which are wired up to publish Bicep use this same manifest writer, but
+        // they all have a callback to work against the IConstruct/BicepGenerationContext. These
+        // will have default implementations but can be overridden if the developer wants to
+        // override how things work for specific scenarios.
+        var annotation = resource.Annotations.OfType<BicepGenerationCallbackAnnotation>().Single();
+        annotation.Callback(bicepGenerationContext);
+
+        // TODO: Pathing is messed up right now. An Infrastructure derived IConstruct seems
+        // to be able to only emit main.bicep where we really should let AZD control stitching
+        // together the various Bicep files we are dropping out.
+        var path = context.GetManifestRelativePath($"{resource.Name}");
```

Comment:
.. we probably don't want to be limited to just generating `main.bicep`

Reply:
This is changing slightly so lets table this one until we get a new alpha out hopefully today or tomorrow.

The general idea is you will get 1 module per (Tenant, ManagementGroup, Subscription, ResourceGroup) that you have added a resource directly to plus one main.bicpe stitching those together with all the param / outputs.

If you did not add any tenant level resources you won't get that module.  If you did not add any subscription level resources you won't get that module etc.

Reply:
Look forward to seeing it. We do really want to be able to emit just the bicep for an individual resource (and any adjacent resources that make sense to be deployed along side it). The Aspire abstract is much more app centric than infrastructure centric and we need the retain the Aspire resource structure in our manifest so that things like connection string wireup can work (connection string injection will remain part of what AZD does when it updates containers etc).

Reply:
I might be missing something but it doesn't appear that any of the bicep examples for Aspire today have a single resource especially when it comes to the "project" concept which is a set of resources with one entry in the manifest.  Other things like [SqlServer](https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting.Azure/Bicep/sql.bicep) appear to have a single bicep file with many resources in it.

I wonder if we have a mismatch on the definition of what a resource is.  From the CDK perspective a resource is anything that can be uniquely queried lets say with a GET operation by a unique set of keys (id).

In this definition SqlDatabase is a resource as well as SqlServer, but in bicep you cannot really separate these two things they need to go in the same module.

By resource from Aspire's terminology do we mean any child of ResourceGroup, Subscription, Tenant, or ManagementGroup?